### PR TITLE
[SYCL] move more common utils from pi_level_zero

### PR DESCRIPTION
**Pull request by igchor**
to UR.

### CODE REVIEW

File: sycl/plugins/unified_runtime/ur/ur.hpp [hunk 1]
```diff
@@ -99,6 +99,101 @@ class SpinLock {
   std::atomic_flag MLock = ATOMIC_FLAG_INIT;
 };
 
+// The wrapper for immutable Level-Zero data.
```
Comment (smaslov-intel) [inline sycl/plugins/unified_runtime/ur/ur.hpp:102]:
```suggestion
// The wrapper for immutable data.

```

File: sycl/plugins/unified_runtime/ur/ur.hpp [hunk 2]
```diff
@@ -99,6 +99,101 @@ class SpinLock {
...
+
+private:
+  std::atomic<pi_uint32> RefCount;
+};
+
+// Base class to store common data
+struct _pi_object {
+  _pi_object() : RefCount{} {}
+
+  // Level Zero doesn't do the reference counting, so we have to do.
```
Comment (smaslov-intel) [inline sycl/plugins/unified_runtime/ur/ur.hpp:178]: Try to get rid of direct mention of Level Zero in "UR"
Also we need to think how this affects backends that do refcnt, e.g. opencl

Comment (igchor) [reply sycl/plugins/unified_runtime/ur/ur.hpp]: Done. Yes, we need to see how useful this `_pi_object` is for other adapters.

Also, we probably need to do something with the env variables, like `SYCL_PI_LEVEL_ZERO_SINGLE_THREAD_MODE`. Should we have separate ones for PI and UR?

**Comment by igchor** [comment]
@smaslov-intel what is the general recommendation for what should stay in pi_level_zero.cpp vs what should be transferred to ur/adapters/level_zero/ur_level_zero.cpp?

Do we want to end up implementing most of the pi_level_zero interfaces using UR, or do we want to have a more independent implementation with just a few functions (like *PlatformGet), utils, and data structures shared?

As long as we don't introduce any overhead, we can probably start moving the implementation of pi_context, pi_device, pi_queue, etc. to the UR, right?

**Comment by smaslov-intel** [comment]
> @smaslov-intel what is the general recommendation for what should stay in pi_level_zero.*pp vs what should be transferred to ur/adapters/level_zero/ur_level_zero.*pp?
> 
> Do we want to end up implementing most of the pi_level_zero interfaces using UR (and just adding interop), or do we want to have a more independent implementation with just a few functions (like *PlatformGet), utils, and data structures shared?
> 
> As long as we don't introduce any overhead, we can probably start moving the implementation of pi_context, pi_device, pi_queue, etc. to the ur/adapters/level_zero/ur_level_zero.*pp, right?

Right, we want to move and share as much as possible.
Ideally, after some short time, we will be able to completely get rid of the pi_level_zero plugin and switch to ur_level_zero plugin, which would be doing the same in the beginning.

File: sycl/plugins/unified_runtime/ur/ur.hpp [hunk 3]
```diff
@@ -99,6 +101,100 @@ class SpinLock {
   std::atomic_flag MLock = ATOMIC_FLAG_INIT;
 };
 
+// The wrapper for immutable data.
+// The data is initialized only once at first access (via ->) with the
+// initialization function provided in Init. All subsequent access to
+// the data just returns the already stored data.
+//
+template <class T> struct ZeCache : private T {
```
Comment (steffenlarsen) [inline sycl/plugins/unified_runtime/ur/ur.hpp:109]: I may be wrong, but `Ze` seems like a leftover from L0-related naming. Should it be renamed to be more general?

**Review by steffenlarsen — APPROVED** [review]
Approving with a small comment. I will leave the final say to @smaslov-intel .

File: sycl/plugins/unified_runtime/ur/ur.hpp (hunk 3 above)
Comment (smaslov-intel) [reply sycl/plugins/unified_runtime/ur/ur.hpp:109]: For sure, but renaming it here means re-naming uses in PI L0 plugin too, which might be big.
I'd prefer us to rename things that are completely moved to UR already.

**Review by smaslov-intel — APPROVED** [review]

**Pull request merged**